core/job/service/job_service.go

```diff
@@ -1999,3 +1993,143 @@ func (j *JobService) getDownstreamJobs(ctx context.Context, projectName tenant.P
 	}
 	return downstreamJobs, me.ToErr()
 }
+
+func (j *JobService) BulkDeleteJobs(ctx context.Context, projectName tenant.ProjectName, jobsToDelete []*dto.JobToDeleteRequest) (map[string]dto.BulkDeleteTracker, error) {
+	// initiate deletion status per jobs to be deleted
+	deletionTrackerMap := map[string]dto.BulkDeleteTracker{}
+
+	toDeleteJobs := job.Jobs{}
+	for _, jobToDelete := range jobsToDelete {
+		existingJob, err := j.jobRepo.GetByJobName(ctx, projectName, jobToDelete.JobName)
+		if err != nil && !errors.IsErrorType(err, errors.ErrNotFound) {
+			j.logger.Error("error getting existing job %s: %s", jobToDelete.JobName.String(), err)
+			return deletionTrackerMap, err
+		}
+
+		toDeleteJobs = append(toDeleteJobs, existingJob)
+	}
+
+	toDeleteMap := toDeleteJobs.GetNameMap()
+	toDeleteSpecMap := toDeleteJobs.GetFullNameToSpecMap()
+
+	alreadyDeleted := map[job.FullName]bool{}
+	deletedJobs := job.Jobs{}
+	for _, toDeleteJob := range toDeleteJobs {
+		deletionTracker, deletedJobNames := j.resolveJobAndDownstreamsDeletion(ctx, toDeleteJob, toDeleteSpecMap, alreadyDeleted)
+		for _, deletedJobName := range deletedJobNames {
+			deletedJob := toDeleteMap[deletedJobName]
+
+			deletedJobs = append(deletedJobs, deletedJob)
+			alreadyDeleted[job.FullName(deletedJob.FullName())] = true
+		}
+
+		for jobName, tracker := range deletionTracker {
+			if _, alreadyDeleted := deletionTrackerMap[jobName]; !alreadyDeleted {
+				deletionTrackerMap[jobName] = tracker
+			}
+		}
+	}
+
+	me := errors.NewMultiError("error cleanup jobs in scheduler")
+	deletedJobsPerNamespace := deletedJobs.GetNamespaceNameAndJobsMap()
+	for namespaceName, deletedJobs := range deletedJobsPerNamespace {
+		tnnt, _ := tenant.NewTenant(projectName.String(), namespaceName.String())
+		err := j.uploadJobs(ctx, tnnt, nil, job.Jobs(deletedJobs).GetJobNames())
+		if err != nil {
+			me.Append(err)
+		}
+	}
+
+	return deletionTrackerMap, me.ToErr()
+}
+
+func (j *JobService) validateDeleteWithDownstreams(ctx context.Context, toDeleteJob *job.Job, toDeleteSpecMap map[job.FullName]*job.Spec) ([]*job.Downstream, error) {
+	downstreamsPerLevel, err := j.getAllDownstreams(ctx, toDeleteJob.ProjectName(), toDeleteJob.Spec().Name(), map[job.FullName]bool{}, 0)
+	if err != nil {
+		return nil, err
+	}
+
+	err = validateDeleteJob(downstreamsPerLevel, toDeleteSpecMap)
+	if err != nil {
+		return nil, err
+	}
+
+	// flatten downstreams per level into a single list,
+	// with direct downstreams will be put first & the leaf downstreams at the end
+	downstreams := []*job.Downstream{}
+	for _, dss := range downstreamsPerLevel {
```

Comment:
can we rename it with more readable variable? eg. `currentDownstreams`

Reply:
agree